src/orange/univ/SimpleUnitTracker.hh

```diff
@@ -546,22 +543,30 @@ SimpleUnitTracker::complex_intersect(LocalState const& state,
     {
         // Index into the distance/face arrays
         size_type const isect = state.temp_next.isect[isect_idx];
+        real_type const distance = state.temp_next.distance[isect];
         // Face being crossed in this ordered intersection
         FaceId face = state.temp_next.face[isect];
-        // Flip the sense of the face being crossed
-        calc_senses.flip_sense(face);
+        Real3 pos = state.pos;
+
+        // cross the internal surface, effectively flipping
+        // internal senses
+        axpy(distance + bump_dist, state.dir, &pos);
```

Comment:
We *do not* want to bump the surface crossings if ever we can (and I don't think we need to here). Since at each intersection we're testing "on that face" (on the other side), I think we should calculate the sense from the previous point (using the previous on_face if we're jumping across the inside of a two-intersection quadric), then flip that sense as we cross it.

Reply:
What about this:

At a point `p`, potentially "on a face" with a face ID `f` and sense `s`:
- Let `s'` be sense wrt the next face `f'` from `p` on `f` (no dynamic storage needed, just calling the lazy sense evaluator)
- Update `p`: `axpy` with the distance to the next face
- Update the current face and flip the sense: `f←f'` and `s←~s'`

Reply:
mmh right, if we need to avoid bumping then a combination of updating on_face and the position should work. Is the reason you avoid bumping numerical instability?

Reply:
Inaccuracy, really: you can jump across corners or nearby surfaces very easily. It turns an "exact" surface crossing into an inexact one.

Reply:
I see, current implementation should be better

Reply:
I made an additional change that correctly implements the algorithm I described above. The version that flips the sense from the *original* position fails if the track moves through a surface twice before exiting.

Maybe it would be good to add little unit test where we have a union of a sphere and a box so that we can test tracking through the double interior crossing without exiting the geometry.

Reply:
the distance is from the original position though? So we'd need to substract the previous distance to get the delta. Also, taking references to the VolumView has caused some issues before where the sense calculator outlives the view.

Reply:
The failing test is why taking references is error-prone. I think the lifetime of the temporary is only extended until the end of the statement, so we end up with a dangling reference as the member. From [cppreference](https://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary):

> a temporary bound to a reference parameter in a function call exists until the end of the full expression containing that function call: if the function returns a reference, which outlives the full expression, it becomes a dangling reference.

Reply:
Yeah I should've checked the unit test. But on GPU you don't want to be glib about copying data locally since that can increase local memory and register usage.

Reply:
Yes sorry, you're quite right I should've updated the distance from the original position